src/renderers/shared/fiber/ReactFiberClassComponent.js

```diff
@@ -70,20 +71,155 @@ module.exports = function(scheduleUpdate : (fiber: Fiber, priorityLevel : Priori
+    // componentWillMount and before this componentWillMount? Probably
+    // unsupported anyway.
+
+    const updateQueue = workInProgress.updateQueue;
+
+    // If this completed, we might be able to just reuse this instance.
+    if (typeof instance.shouldComponentUpdate === 'function' &&
+        !(updateQueue && updateQueue.isForced) &&
+        workInProgress.memoizedProps !== null &&
+        !instance.shouldComponentUpdate(newProps, newState)) {
+      return false;
+    }
+
+    // If we didn't bail out we need to construct a new instance. We don't
+    // want to reuse one that failed to fully mount.
+    const newInstance = constructClassInstance(workInProgress);
+    newInstance.props = newProps;
+    newInstance.state = newState = newInstance.state || null;
+
+    if (typeof newInstance.componentWillMount === 'function') {
+      newInstance.componentWillMount();
+      // If we had additional state updates during this life-cycle, let's
+      // process them now.
+      const newUpdateQueue = workInProgress.updateQueue;
+      if (newUpdateQueue) {
+        instance.state = mergeUpdateQueue(newUpdateQueue, newState, newProps);
+      }
+    }
+    return true;
+  }
+
+  // Invokes the update life-cycles and returns false if it shouldn't rerender.
+  function updateClassInstance(current : Fiber, workInProgress : Fiber) : boolean {
+    const instance = workInProgress.stateNode;
+
+    const oldProps = current.memoizedProps;
+    let newProps = workInProgress.pendingProps;
+    if (!newProps) {
+      // If there aren't any new props, then we'll reuse the memoized props.
+      // This could be from already completed work.
+      newProps = workInProgress.memoizedProps;
+      if (!newProps) {
+        throw new Error('There should always be pending or memoized props.');
+      }
+    }
+
+    // Note: During these life-cycles, instance.props/instance.state are what
+    // ever the previously attempted to render - not the "current". However,
+    // during componentDidUpdate we pass the "current" props.
+
+    if (oldProps !== newProps) {
+      if (typeof instance.componentWillReceiveProps === 'function') {
+        instance.componentWillReceiveProps(newProps);
+      }
+    }
+
+    // Compute the next state using the memoized state and the update queue.
+    const updateQueue = workInProgress.updateQueue;
+    const previousState = workInProgress.memoizedState;
+    // TODO: Previous state can be null.
+    let newState;
+    if (updateQueue) {
+      newState = mergeUpdateQueue(updateQueue, previousState, newProps);
+    } else {
+      newState = previousState;
+    }
+
+    if (typeof instance.shouldComponentUpdate === 'function' &&
+        !(updateQueue && updateQueue.isForced) &&
+        workInProgress.memoizedProps !== null &&
+        !instance.shouldComponentUpdate(newProps, newState)) {
+      // TODO: Should this get the new props/state updated regardless?
```

Comment:
Do you mean whether we should reassign `this.props` and `this.state` despite sCU returning `false`? Is there any reason not to?